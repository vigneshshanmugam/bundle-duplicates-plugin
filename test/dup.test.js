const path = require("path");
const fs = require("fs");
const webpack = require("webpack");
const rimraf = require("rimraf");
const { CPA, stringify } = require("js-cpa");
const distDir = path.join(__dirname, "dist");
const fixtures = path.join(__dirname, "fixtures");

const runWebpack = (config, callback) => {
  const compiler = webpack(config);
  compiler.run((err, stats) => {
    if (err) {
      console.error(err);
      return;
    }
    const compilation = stats.compilation;
    // Same as the webpack plugin
    // Much easier to test than mocking process.stdout and filename path
    const cpa = new CPA({ threshold: 1 });
    Object.keys(compilation.assets).forEach(filename => {
      cpa.add(compilation.assets[filename].source(), {
        filename
      });
    });

    callback(cpa.findOptimalDuplicates());
  });
};

const simpleConfig = {
  entry: path.join(fixtures, "simple", "entry.js"),
  output: {
    path: distDir,
    filename: "bundle.js"
  }
};

const commonChunksConfig = {
  entry: {
    page1: path.join(fixtures, "common-chunk", "page-1"),
    page2: path.join(fixtures, "common-chunk", "page-2")
  },
  output: {
    path: distDir,
    filename: "[name].js"
  },
  plugins: [
    new webpack.optimize.CommonsChunkPlugin({
      name: "commons",
      chunks: ["page1", "page2"]
    })
  ]
};

afterEach(() => {
  rimraf.sync(distDir);
});

test("simple - should output duplicates", done => {
  return runWebpack(simpleConfig, duplicates => {
    expect(
      stringify(duplicates, {
        colors: false,
        newline: false
      })
    ).toMatchSnapshot();
    done();
  });
});

test("common chunks - should output duplicates", done => {
  return runWebpack(commonChunksConfig, duplicates => {
    expect(
      stringify(duplicates, {
        colors: false,
        newline: false
      })
    ).toMatchSnapshot();
    done();
  });
});
