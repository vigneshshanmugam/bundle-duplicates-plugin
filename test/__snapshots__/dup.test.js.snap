// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common chunks - should output duplicates 1`] = `
"Found 1 matches

Duplicate 1

page2.js:10:12
function page2() {
  const b = 20;
}

page1.js:8:10
function page1() {
  const a = 20;
}

commons.js:154:156
function common() {
  const c = 20;
}

"
`;

exports[`simple - should output duplicates 1`] = `
"Found 1 matches

Duplicate 1

bundle.js:80:86
function foo() {
  const a = 10;
  function test1() {
    return function g1() {};
  }
  return \\"124\\";
}

bundle.js:97:103
function foo() {
  const a = 10;
  function test2() {
    return function g2() {};
  }
  return \\"124\\";
}

"
`;
